### PR TITLE
Plugins v3

### PR DESCRIPTION
**Pull request by catornot**
yes

launcher pr : https://github.com/R2Northstar/NorthstarLauncher/pull/472

### CODE REVIEW

**Comment by ASpoonPlaysGames** [comment]
Would it be feasible to remove the launcher -> Northstar.Client dependency by reworking presence? I think ideally it gets handled all in launcher, but if not I would prefer it gets split off into a different mod or something?

**Comment by catornot** [comment]
> Would it be feasible to remove the launcher -> Northstar.Client dependency by reworking presence?

launcher doesn't handle presence in v3, I want to move it to the drpc plugin. 

> I think ideally it gets handled all in launcher, but if not I would prefer it gets split off into a different mod or something?

It might be possible but some stuff is just easier to access on the sqvm.

**Review by ASpoonPlaysGames — APPROVED** [review]
Code looks good, though I would prefer if Launcher didn't depend on Mods.

Then again, it already does for a lot of stuff so ehhhh

**Comment by catornot** [comment]
> though I would prefer if Launcher didn't depend on Mods.

The discord rpc plugin depends on this now. Launcher doesn't handle presence anymore.

**Comment by ASpoonPlaysGames** [comment]
> The discord rpc plugin depends on this now. Launcher doesn't handle presence anymore.

Perhaps rename the `NorthstarCodeCallback` functions then? Since they aren't coming from Northstar itself but a plugin. Something like `DiscordRPC_` as a prefix would work imo

File: Northstar.Client/mod/scripts/vscripts/presence/ui_presence.nut [hunk 1]
```diff
@@ -1,12 +1,15 @@
...
-	UIPresenceStruct uis
+UIPresenceStruct function DiscordRPC_GenerateUIPresence( UIPresenceStruct uis ) {
+	if (uiGlobal.isLoading)
+		uis.game_state = eDiscordGameState.LOADING;
+	else if (uiGlobal.loadedLevel == "")
+		uis.game_state = eDiscordGameState.MAINMENU;
+	else if (IsLobby() || uiGlobal.loadedLevel == "mp_lobby")
+		uis.game_state = eDiscordGameState.LOBBY;
+	else
+		uis.game_state = eDiscordGameState.INGAME;
```
Comment (ASpoonPlaysGames) [inline Northstar.Client/mod/scripts/vscripts/presence/ui_presence.nut:12]:
```suggestion
UIPresenceStruct function DiscordRPC_GenerateUIPresence( UIPresenceStruct uis )
{
	if ( uiGlobal.isLoading )
		uis.game_state = eDiscordGameState.LOADING;
	else if ( uiGlobal.loadedLevel == "" )
		uis.game_state = eDiscordGameState.MAINMENU;
	else if ( IsLobby() || uiGlobal.loadedLevel == "mp_lobby" )
		uis.game_state = eDiscordGameState.LOBBY;
	else
		uis.game_state = eDiscordGameState.INGAME;
```

File: Northstar.Client/mod/scripts/vscripts/presence/cl_presence.nut [hunk 2]
```diff
@@ -40,6 +37,6 @@ void function NorthstarCodeCallback_GenerateGameState() {
 		gs.timeEnd = expect float(level.nv.roundEndTime - Time())
 	else
 		gs.timeEnd = expect float(level.nv.gameEndTime - Time())
-
-    NSPushGameStateData(gs)
+
```
Comment (ASpoonPlaysGames) [inline Northstar.Client/mod/scripts/vscripts/presence/cl_presence.nut:40]:
```suggestion

```
whitespace

File: Northstar.Client/mod/scripts/vscripts/presence/cl_presence.nut [hunk 3]
```diff
@@ -1,10 +1,7 @@
 untyped
 globalize_all_functions
 
-void function NorthstarCodeCallback_GenerateGameState() {
-
-    GameStateStruct gs
-
+GameStateStruct function DiscordRPC_GenerateGameState( GameStateStruct gs ) {
```
Comment (ASpoonPlaysGames) [inline Northstar.Client/mod/scripts/vscripts/presence/cl_presence.nut:4]:
```suggestion
GameStateStruct function DiscordRPC_GenerateGameState( GameStateStruct gs )
{
```

**Review by ASpoonPlaysGames — CHANGES_REQUESTED** [review]
Formatting stuff

**Comment by ASpoonPlaysGames** [comment]
Just going to mention that afaik this PR does not depend directly on the Launcher PR (https://github.com/R2Northstar/NorthstarLauncher/pull/472) but instead depends on https://github.com/R2Northstar/NorthstarDiscordRPC/pull/10 (which does depend on https://github.com/R2Northstar/NorthstarLauncher/pull/472)

Basically what I'm getting at is that this PR shouldn't block https://github.com/R2Northstar/NorthstarLauncher/pull/472 but also cannot properly be tested without https://github.com/R2Northstar/NorthstarLauncher/pull/472 and https://github.com/R2Northstar/NorthstarDiscordRPC/pull/10

File: .github/nativefuncs.json [hunk 4]
```diff
@@ -723,18 +717,6 @@
...
-			"helpText":"",
-			"returnTypeString":"void",
-			"argTypes":"struct presence"
-		},
-		{
-			"name":"NSGetMasterServerAuthResult",
-			"helpText":"",
-			"returnTypeString":"MasterServerAuthResult",
-			"argTypes":""
-		}
```
Comment (ASpoonPlaysGames) [inline .github/nativefuncs.json:738]: accidental removal? causing compile check failure

Comment (catornot) [reply .github/nativefuncs.json:738]: yeah, I wanted to fix it but didn't have time to.

File: Northstar.Client/mod/scripts/vscripts/cl_northstar_client_init.nut [hunk 5]
```diff
@@ -15,10 +23,7 @@ global struct GameStateStruct {
 }
 
 global struct UIPresenceStruct {
-	bool isLoading
-	bool isLobby
-	string loadingLevel
-	string loadedLevel
+	int game_state
```
Comment (ASpoonPlaysGames) [inline Northstar.Client/mod/scripts/vscripts/cl_northstar_client_init.nut:26]: Should be camelCase

File: Northstar.Client/mod/scripts/vscripts/cl_northstar_client_init.nut [hunk 6]
```diff
@@ -1,3 +1,11 @@
+global enum eDiscordGameState
+{
+	LOADING = 0
+	MAINMENU
+    LOBBY
+    INGAME
+}
```
Comment (ASpoonPlaysGames) [inline Northstar.Client/mod/scripts/vscripts/cl_northstar_client_init.nut:7]: Weird whitespace going on here, mixed tabs and spaces probably

File: Northstar.Client/mod/scripts/vscripts/presence/ui_presence.nut [hunk 7]
```diff
@@ -1,12 +1,16 @@
...
 
-	uis.isLoading = uiGlobal.isLoading
-	uis.isLobby = IsLobby()
-	uis.loadingLevel = uiGlobal.loadingLevel
-	uis.loadedLevel = uiGlobal.loadedLevel
-	NSPushUIPresence(uis)
+	return uis
 }
```
Comment (ASpoonPlaysGames) [inline Northstar.Client/mod/scripts/vscripts/presence/ui_presence.nut:16]:
```suggestion
}

```
Newline at end of file

File: Northstar.Client/mod/scripts/vscripts/presence/cl_presence.nut [hunk 8]
```diff
@@ -40,6 +38,5 @@ void function NorthstarCodeCallback_GenerateGameState() {
 		gs.timeEnd = expect float(level.nv.roundEndTime - Time())
 	else
 		gs.timeEnd = expect float(level.nv.gameEndTime - Time())
-
-    NSPushGameStateData(gs)
+    return gs
 }
```
Comment (ASpoonPlaysGames) [inline Northstar.Client/mod/scripts/vscripts/presence/cl_presence.nut:42]:
```suggestion
}

```
Newline at end of file

**Review by ASpoonPlaysGames — CHANGES_REQUESTED** [review]
Just some formatting

**Review by ASpoonPlaysGames — APPROVED** [review]
Code looks good.

I'm ignoring the weird whitespace in some files because the whole file is wrong, so fixing is beyond the scope of this PR

**Review by Jan200101 — APPROVED** [review]
Tested as part of the Launcher PR

**Review by GeckoEidechse — APPROVED** [review]
Confirmed working in testing together with https://github.com/R2Northstar/NorthstarLauncher/pull/472 and https://github.com/R2Northstar/NorthstarDiscordRPC/pull/10

Joined a bunch of servers and watched Discord activity status update

(No code review done in this review)

**Pull request merged**